pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,255 @@
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static void initThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get(); // Just make sure it's initialized
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().size() == 1) {
+            finishOperation();
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     */
+    public static void startOperation(final TimedOperationCategory category) {
+        startOperation(category, null);
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     * @param label A label to be added to the category. Allows to differentiate measures within a single category.
+     */
+    public static void startOperation(final TimedOperationCategory category, final String label) {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get().add(new TimerEntry(category, label));
+    }
+    
+    /**
+     * Finishes tracking an operation.
+     */
+    public static void finishOperation() {
+        finishOperation(0);
+    }
+    
+    /**
+     * Finishes tracking an operation.
+     * @param extraDataCounter An optional additional data counter to track along the measurements.
+     *                         Users are free to track any extra value they want (ie: number of analyzed nodes,
+     *                         iterations in a loop, etc.)
+     */
+    public static void finishOperation(final long extraDataCounter) {
+        if (!trackTime) {
+            return;
+        }
+
+        final Queue<TimerEntry> queue = TIMER_ENTRIES.get();
+        final TimerEntry timerEntry = queue.remove();
+
+        // Compute if absent
+        TimedResult result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        if (result == null) {
+            ACCUMULATED_RESULTS.putIfAbsent(timerEntry.operation, new TimedResult());
+            result = ACCUMULATED_RESULTS.get(timerEntry.operation);
+        }
+
+        // Update counters and let next element on the stack ignore the time we spent
+        final long delta = result.accumulate(timerEntry, extraDataCounter);
+        if (!queue.isEmpty()) { 
+            queue.peek().inNestedOperations += delta;
+        }
+    }
+    
+    /**
+     * An entry in the open timers queue. Defines an operation that has started and hasn't finished yet.
+     */
+    private static class TimerEntry {
+        /* package */ final TimedOperation operation;
+        /* package */ final long start;
+        /* package */ long inNestedOperations = 0;
```

Comment:
Hmm, this might throw a wrench into my thinking so far.  I was optimistic the thread-local could go away in my other comments on a `Tracker` try-with-resources.  There's no way to reach across the Java stack without using a thread local though.

It's possible to not use a Queue and have a parent pointer on the `Tracker` instances, then the thread-local only need point to the nearest nested `Tracker` instance.  Probably not worth the complexity, although perhaps it could reduce object allocations around queue management.